repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -245,4 +263,24 @@ class Session(
     // Set the job group
     executeCode(statementId, cmd)
   }
+
+  private def cleanOldStatements(): Unit = {
+    if (_statements.size > numRetainedStatements) {
+      // Remove 10% of existing statements to avoid frequently calling this method when reaching
+      // threshold.
+      val toRemove = (numRetainedStatements * 0.1).toInt + 1
+      val stmtIdToRemove = math.max(newStatementId.get() - _statements.size, 0)
+      (stmtIdToRemove until stmtIdToRemove + toRemove).foreach { id =>
+        _statements.get(id).foreach { st =>
```

Comment:
It may not be so expensive as hashmap operation is `O(1)`, so the total cost would be `n * O(1)`. Yes `LinkedHashMap` is a better solution for this, let me try to solve this.